### PR TITLE
feat: added httpsignature check to /hook endpoint via DRONE_SECRET

### PR DESCRIPTION
**Pull request by davideperozzi**
This adds a basic http signature verification to the setup. I'm not an expert with python, so you might want to check the code. 

I added the implementation from here: https://datatracker.ietf.org/doc/html/draft-cavage-http-signatures-10 as mentioned here https://docs.drone.io/extensions/secret/#authorization. Only did the hmac implementation, because Drone doesn't use RSA (does it?). 

I added a unit test and a script to mimic the behavior of the drone image without the need to setup a drone instance etc. (would be probably better for testing tho, because of changes). I added it to the readme how you can verify it. 

So basically you need to define the "DRONE_SECRET" env var in order to access the service. It's not perfect, but it does the job. Feel free to add comments to the lines!

This should close this issue, when it's been integrated properly: https://github.com/spritsail/drone-notify/issues/9

Edit: This forces a new major version, if we make it mandatory. We could also make the DRONE_SECRET option, which I wouldn't recommend.

### CODE REVIEW

**Comment by frebib** [comment]
Thanks for the PR! Nice work on the tests too!
There are a few bits I'd like to change before integrating if that's okay. I'll push some changes to this PR so you can review and when we're both happy I can squash and merge 👍

~Right now my main question is: you only seem to be including the headers in the signature generation and not the request body? That doesn't seem too useful 🤔 Does Drone not include the request body as part of it's signature?~
_Edit_: It looks like Drone signs all headers and the request body. Our code will need to handle that: https://github.com/harness/gitness/blob/drone/plugin/webhook/webhook.go#L86-L97

**Comment by davideperozzi** [comment]
@frebib Sounds good! Yeah I thought that too, but it seems that it only includes the headers in the signature. https://github.com/99designs/httpsignatures-go/blob/master/signer.go#L63-L78. I'm not a security expert, but I think without the body in the signature, it's prone to MITM, since you can just intercept the request and change the body? 

I couldn't find the place in the httpsignatures-go package where they include the body in the hash. Do you know where they do it?

**Comment by frebib** [comment]
Sorry for the radio silence here. It's been a busy couple of weeks.
I found some time today to work on this, and I took what you did and changed it up a little. I also added functionality to validate the `Digest:` header so we know that the signature is covering the request body. If you'd like to have a look, the commits are here https://github.com/spritsail/drone-notify/commits/httpsignature/
I pushed this to Docker Hub as `spritsail/drone-notify:httpsignature`
I would like to spend some time reworking your tests into pytest and adding some more before merging this, but until I do that please feel free to use that branch and if you're so inclined send any PRs to adjust it as you see fit.
Thanks for your work here. It made my refactoring a lot easier 😁

**Comment by frebib** [comment]
Merged into master in 02c6bf1
Thanks for your work here!
Let me know if you have any issues with this

**Comment by davideperozzi** [comment]
No worries, thank you - happy to help! I'll try it by the end of the week, looks good:)
